packages/web3-utils/src/socket_provider.ts

```diff
@@ -197,28 +199,88 @@ export abstract class SocketProvider<
 	 */
-	public on<T = JsonRpcResult>(type: EventType, callback: Web3ProviderEventCallback<T>): void {
-		this._eventEmitter.on(type, callback);
+	public on(
+		type: 'disconnect',
+		listener: Web3Eip1193ProviderEventCallback<ProviderRpcError>,
+	): void;
+	public on(
+		type: 'connect',
+		listener: Web3Eip1193ProviderEventCallback<ProviderConnectInfo>,
+	): void;
+	public on(type: 'chainChanged', listener: Web3Eip1193ProviderEventCallback<string>): void;
+	public on(type: 'accountsChanged', listener: Web3Eip1193ProviderEventCallback<string[]>): void;
+	public on<T = JsonRpcResult>(
+		type: 'message' | string,
+		listener: Web3Eip1193ProviderEventCallback<ProviderMessage> | Web3ProviderEventCallback<T>,
+	): void;
+	public on<T = JsonRpcResult>(
+		type: string,
+		listener: Web3Eip1193ProviderEventCallback<any> | Web3ProviderEventCallback<T>,
+	): void {
+		this._eventEmitter.on(type, listener);
 	}
 
 	/**
 	 * Registers a listener for the specified event type that will be invoked at most once.
 	 * @param type  - The event type to listen for
 	 * @param callback - The callback to be invoked when the event is emitted
 	 */
-	public once<T = JsonRpcResult>(type: EventType, callback: Web3ProviderEventCallback<T>): void {
-		this._eventEmitter.once(type, callback);
+	public once(
+		type: 'disconnect',
+		listener: Web3Eip1193ProviderEventCallback<ProviderRpcError>,
+	): void;
+	public once(
+		type: 'connect',
+		listener: Web3Eip1193ProviderEventCallback<ProviderConnectInfo>,
+	): void;
+	public once(type: 'chainChanged', listener: Web3Eip1193ProviderEventCallback<string>): void;
+	public once(
+		type: 'accountsChanged',
+		listener: Web3Eip1193ProviderEventCallback<string[]>,
+	): void;
+	public once<T = JsonRpcResult>(
+		type: 'message' | string,
+		listener: Web3Eip1193ProviderEventCallback<ProviderMessage> | Web3ProviderEventCallback<T>,
+	): void;
+	public once<T = JsonRpcResult>(
+		type: string,
+		listener: Web3Eip1193ProviderEventCallback<any> | Web3ProviderEventCallback<T>,
+	): void {
+		this._eventEmitter.once(type, listener);
 	}
 
 	/**
 	 *  Removes a listener for the specified event type.
 	 * @param type - The event type to remove the listener for
 	 * @param callback - The callback to be executed
 	 */
-	public removeListener(type: EventType, callback: Web3ProviderEventCallback): void {
-		this._eventEmitter.removeListener(type, callback);
+	public removeListener(
+		type: 'disconnect',
+		listener: Web3Eip1193ProviderEventCallback<ProviderRpcError>,
+	): void;
+	public removeListener(
+		type: 'connect',
+		listener: Web3Eip1193ProviderEventCallback<ProviderConnectInfo>,
+	): void;
+	public removeListener(
+		type: 'chainChanged',
+		listener: Web3Eip1193ProviderEventCallback<string>,
+	): void;
+	public removeListener(
+		type: 'accountsChanged',
+		listener: Web3Eip1193ProviderEventCallback<string[]>,
+	): void;
+	public removeListener<T = JsonRpcResult>(
+		type: 'message' | string,
```

Comment:
type: 'message' | string,
I think it should be only 'message', or is there reason of allowing it string as well?